File: core/src/main/java/pcb/uwu/units/composite/termodynamics/PowerUnit.java
package pcb.uwu.units.composite.termodynamics;

import pcb.uwu.core.CompositeUnit;
import pcb.uwu.core.UnitCounter;
import pcb.uwu.units.composite.electromagnetism.ElectricPotentialUnit;
import pcb.uwu.units.composite.mechanics.ForceUnit;
import pcb.uwu.units.quantity.ElectricCurrentUnit;
import pcb.uwu.units.quantity.LengthUnit;
import pcb.uwu.units.quantity.MassUnit;
import pcb.uwu.units.quantity.TimeUnit;

public class PowerUnit extends CompositeUnit {

	// region base constructor

	public PowerUnit(LengthUnit lengthUnit, MassUnit massUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(massUnit)
				.major(lengthUnit, 2)
				.minor(timeUnit, 3));
	}

	// endregion

	// region composite constructors

	public PowerUnit(ForceUnit forceUnit, LengthUnit lengthUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(forceUnit.getUnitCounter())
				.major(lengthUnit)
				.minor(timeUnit));
	}

	public PowerUnit(EnergyUnit energyUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(energyUnit.getUnitCounter())
				.minor(timeUnit));
	}

	public PowerUnit(PowerUnit powerUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(powerUnit.getUnitCounter())
				.major(timeUnit));
	}

	public PowerUnit(ElectricCurrentUnit electricCurrentUnit, ElectricPotentialUnit electricPotentialUnit) {
		super(new UnitCounter()
				.major(electricCurrentUnit.getUnitCounter())
				.major(electricPotentialUnit.getUnitCounter()));
	}

	// endregion

}

File: dimension/src/main/java/pcb/uwu/unit/derived/termodynamics/PowerUnit.kt
package pcb.uwu.unit.derived.termodynamics

import pcb.uwu.core.CompositeUnit
import pcb.uwu.core.UnitCounter
import pcb.uwu.unit.base.ElectricCurrentUnit
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.MassUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.electromagnetism.ElectricPotentialUnit
import pcb.uwu.unit.derived.mechanics.ForceUnit

open class PowerUnit : CompositeUnit
{
    constructor(lengthUnit: LengthUnit,
                massUnit: MassUnit,
                timeUnit: TimeUnit)
            : super(UnitCounter()
                        .major(massUnit)
                        .major(lengthUnit)
                        .major(lengthUnit)
                        .minor(timeUnit)
                        .minor(timeUnit)
                        .minor(timeUnit))

    constructor(forceUnit: ForceUnit,
                lengthUnit: LengthUnit,
                timeUnit: TimeUnit)
            : super(UnitCounter()
                        .major(forceUnit)
                        .major(lengthUnit)
                        .minor(timeUnit))

    constructor(energyUnit: EnergyUnit,
                timeUnit: TimeUnit)
            : super(UnitCounter()
                        .major(energyUnit)
                        .minor(timeUnit))

    constructor(electricCurrentUnit: ElectricCurrentUnit,
                electricPotentialUnit: ElectricPotentialUnit)
            : super(UnitCounter()
                        .major(electricCurrentUnit)
                        .major(electricPotentialUnit))
}
